csrc/python_frontend/segmentation.cpp

```diff
@@ -79,6 +79,167 @@ int64_t SegmentationState::setupSegmentation(
+
+  // Step 2) Create CPP Fusion for SegmentedGroup. The IrCloner acts as a map
+  // from fusion segment to the original fusion.
+  std::pair<IrCloner, std::unique_ptr<Fusion>> cloner_segment_pair =
+      segmented_fusion_->makeFusion(sg);
+  IrCloner cloned_to_segment_map = cloner_segment_pair.first;
+  std::unique_ptr<Fusion> fusion_segment =
+      std::move(cloner_segment_pair.second);
+
+  // Step 3) Translate CPP Fusion to Python FusionDefinition
+  std::unordered_map<const nvfuser::Val*, size_t>
+      map_translated_val_to_segment_fid =
+          translate(fusion_segment.get(), &other);
+
+  // Step 4) Create map from segment fusion indices to original fusion indices.
+  // Step 4a) Get FusionDefinition index for cloned inputs and outputs. Map them
+  // to their original fusion indices.
+  const std::vector<Val*>& cloned_inputs = sg->inputs();
+  const std::vector<Val*>& cloned_outputs = sg->outputs();
+
+  std::vector<int64_t> original_fid;
+  original_fid.reserve(cloned_inputs.size() + cloned_outputs.size());
+
+  std::transform(
+      cloned_inputs.begin(),
+      cloned_inputs.end(),
+      std::back_inserter(original_fid),
+      [&](Val* v) { return map_cloned_value_to_fid_.at(v); });
+
+  std::transform(
+      cloned_outputs.begin(),
+      cloned_outputs.end(),
+      std::back_inserter(original_fid),
+      [&](Val* v) { return map_cloned_value_to_fid_.at(v); });
+
+  // Step 4b) ir_cloner maps cloned fusion Vals to segment Vals.
+  std::vector<Val*> segment_inputs_outputs;
+  segment_inputs_outputs.reserve(cloned_inputs.size() + cloned_outputs.size());
+
+  std::transform(
+      cloned_inputs.begin(),
+      cloned_inputs.end(),
+      std::back_inserter(segment_inputs_outputs),
+      [&](Val* v) { return cloned_to_segment_map.clone(v); });
+
+  std::transform(
+      cloned_outputs.begin(),
+      cloned_outputs.end(),
+      std::back_inserter(segment_inputs_outputs),
+      [&](Val* v) { return cloned_to_segment_map.clone(v); });
+
+  // Step 4c) Map segment Vals to their FusionDefinition index.
+  std::vector<int64_t> segment_fid;
+  segment_fid.reserve(segment_inputs_outputs.size());
+  std::transform(
+      segment_inputs_outputs.begin(),
+      segment_inputs_outputs.end(),
+      std::back_inserter(segment_fid),
+      [&](Val* v) { return map_translated_val_to_segment_fid.at(v); });
+
+  // Step 4d) Map original indices to segment indices.
+  NVF_ERROR(original_fid.size() == segment_fid.size());
+  std::unordered_map<int64_t, int64_t> segment_fid_to_original_fid_map;
+  for (size_t idx : c10::irange(original_fid.size())) {
+    segment_fid_to_original_fid_map.emplace(
+        segment_fid.at(idx), original_fid.at(idx));
+  }
+
+  // Step 4e) short-circuit: No extra extents required for python definition.
+  if (fusion_segment->inputs().size() == other.inputs().size()) {
+    return segment_fid_to_original_fid_map;
+  }
+
+  // The python segment can require the size of tensor dimensions from original
```

Comment:
I have to say that I am quite lost on how the mapping were done in the following section.

iiuc, there are four things that we are trying to map.

1. original Fusion.
2. Fusion segment.
3. original FusionDefinition.
4. FusionDefinition for the segment that we populate in this function.

I find that the use of `translated_xxx` and `cloned_xxx` in the variable name a bit confusing (this could just be me).

Reply:
The mapping is from state indices for the segment to state indices of the original fusion.
e.g., From number 4 to number 3.

`translated_xxx` corresponds with the segment CPP fusion. `cloned_xxx` corresponds with the clone of the original CPP fusion.